src/linear-types/Control/Distributed/Spark/Safe/Closure.hs

```diff
@@ -0,0 +1,303 @@
+            -- send other exceptions in string form
+            Nothing -> do
+              jt <- reflect (Text.pack $ show e)
+              je <- new jt
+              Foreign.JNI.throw (je :: J ('Class "java/lang/RuntimeException"))
+              return jnull
+      )
+
+foreign export ccall "sparkle_apply" apply
+  :: Ptr JByteArray
+  -> Ptr JObjectArray
+  -> IO (Ptr JObject)
+  -}
+
+
+pairDict :: Dict c1 -> Dict c2 -> Dict (c1, c2)
+pairDict Dict Dict = Dict
+
+closFun1
+  :: forall a b.
+    Dict (Reify a, Reflect b)
+     -> (a -> b)
+     -> JObjectArray
+  %1 -> LIO.IO JObject
+closFun1 Dict f args = Control.Functor.Linear.do
+  (args', a) <- S.getObjectArrayElement args 0
+  (objptr, Ur a') <- reif (unsafeCast a)
+  S.deleteLocalRef objptr
+  S.deleteLocalRef args'
+  upcast <$> refl (f a')
+    -- fmap upcast . refl =<< return . f =<< reif . unsafeCast =<< S.getObjectArrayElement args 0
+  where
+    reif = reify :: J (Interp a) %1 -> LIO.IO (J (Interp a), Ur a)
+    refl = reflect :: b -> LIO.IO (J (Interp b))
+
+tripleDict :: Dict c1 -> Dict c2 -> Dict c3 -> Dict (c1, c2, c3)
+tripleDict Dict Dict Dict = Dict
+
+closFun2
+  :: forall a b c.
+    Dict (Reify a, Reify b, Reflect c)
+     -> (a -> b -> c)
+     -> JObjectArray
+  %1 -> LIO.IO JObject
+closFun2 Dict f args = Control.Functor.Linear.do
+    (args', a) <- second unsafeCast <$> S.getObjectArrayElement args 0
+    (args'', b) <- second unsafeCast <$> S.getObjectArrayElement args' 1
+    (aptr, Ur a') <- reifA a
+    (bptr, Ur b') <- reifB b
+    S.deleteLocalRef aptr
+    S.deleteLocalRef bptr
+    S.deleteLocalRef args''
+    upcast <$> reflC (f a' b')
+  where
+    reifA = reify :: J (Interp a) %1 -> LIO.IO (J (Interp a), Ur a)
+    reifB = reify :: J (Interp b) %1 -> LIO.IO (J (Interp b), Ur b)
+    reflC = reflect :: c -> LIO.IO (J (Interp c))
+    second :: (n %1 -> o) -> (m, n) %1 -> (m, o)
+    second g (a, b) = (a, g b) 
+
+clos2bs :: Typeable a => Closure a %1-> ByteString
+clos2bs = Unsafe.toLinear $ LBS.toStrict Prelude.. encode
+
+bs2clos :: Typeable a => ByteString %1-> Closure a
+bs2clos = Unsafe.toLinear $ decode Prelude.. LBS.fromStrict
+
+-- TODO: These two instances should probably go in `jvm` instead of here
+instance (Static (JUnsafe.Reify a), Reify a) => Static (Reify a) where
+  closureDict = closureDict
+
+instance (Static (JUnsafe.Reflect a), Reflect a) => Static (Reflect a) where
+  closureDict = closureDict
+
+-- | Like 'Interp', but parameterized by the target arity of the 'Fun' instance.
+type family InterpWithArity (n :: Nat) (a :: Type) :: JType
+
+-- | A @ReifyFun n a@ constraint states that @a@ is a function type of arity
+-- @n@. That is, a value of this function type can be made from a @JFun{n}@.
+class ReifyFun n a where
+  -- | Like 'reify', but specialized to reifying functions at a given arity.
+  reifyFun :: MonadIO m => Sing n -> J (InterpWithArity n a) %1 -> m (J (InterpWithArity n a), Ur a)
```

Comment:
I made it this way so that it would be analogous to the safe `reify`, which also returns the reference, but I'll add a `reifyFun_` that doesn't in analogy to `reify_`.